bgpd/bgp_nexthop.c

```diff
@@ -813,97 +896,168 @@ static void bgp_show_nexthops_detail(struct vty *vty, struct bgp *bgp,
-		if (!CHECK_FLAG(bnc->flags, BGP_NEXTHOP_REGISTERED))
-			vty_out(vty, "  Is not Registered\n");
+		if (json) {
+			json_object_string_add(json, "ip", buf);
+			json_object_boolean_false_add(json, "valid");
+			json_object_boolean_false_add(json, "complete");
+			json_object_int_add(json, "pathCount", bnc->path_count);
+			if (peer)
+				json_object_string_add(json, "hostname",
+						       peer->host);
+			if (bnc->is_evpn_gwip_nexthop)
+				json_object_boolean_true_add(json,
+							     "isEvpnGatewayIp");
+			if (CHECK_FLAG(bnc->flags, BGP_NEXTHOP_CONNECTED))
+				json_object_boolean_false_add(json,
+							      "isConnected");
+			if (!CHECK_FLAG(bnc->flags, BGP_NEXTHOP_REGISTERED))
+				json_object_boolean_false_add(json,
+							      "isRegistered");
+		} else {
+			vty_out(vty, " %s invalid, #paths %d", buf,
+				bnc->path_count);
+			if (peer)
+				vty_out(vty, ", peer %s", peer->host);
+			if (bnc->is_evpn_gwip_nexthop)
+				vty_out(vty, " EVPN Gateway IP");
+			vty_out(vty, "\n");
+			if (CHECK_FLAG(bnc->flags, BGP_NEXTHOP_CONNECTED))
+				vty_out(vty, "  Must be Connected\n");
+			if (!CHECK_FLAG(bnc->flags, BGP_NEXTHOP_REGISTERED))
+				vty_out(vty, "  Is not Registered\n");
+		}
 	}
 	tbuf = time(NULL) - (monotime(NULL) - bnc->last_update);
-	vty_out(vty, "  Last update: %s", ctime(&tbuf));
+	if (json) {
+		json_last_update = json_object_new_object();
+		json_object_int_add(json_last_update, "epoch", tbuf);
+		json_object_string_add(json_last_update, "string",
+				       ctime(&tbuf));
+		json_object_object_add(json, "lastUpdate", json_last_update);
+	} else {
+		vty_out(vty, "  Last update: %s", ctime(&tbuf));
+	}
 
 	/* show paths dependent on nexthop, if needed. */
 	if (specific)
-		bgp_show_nexthop_paths(vty, bgp, bnc);
+		bgp_show_nexthop_paths(vty, bgp, bnc, json);
 }
 
 static void bgp_show_nexthops(struct vty *vty, struct bgp *bgp,
-			      bool import_table)
+			      bool import_table, json_object *json)
 {
 	struct bgp_nexthop_cache *bnc;
 	afi_t afi;
 	struct bgp_nexthop_cache_head(*tree)[AFI_MAX];
+	json_object *json_nexthops = NULL;
+	json_object *json_nexthop = NULL;
 
-	if (import_table)
-		vty_out(vty, "Current BGP import check cache:\n");
-	else
-		vty_out(vty, "Current BGP nexthop cache:\n");
+	if (!json) {
+		if (import_table)
+			vty_out(vty, "Current BGP import check cache:\n");
+		else
+			vty_out(vty, "Current BGP nexthop cache:\n");
+	}
 	if (import_table)
 		tree = &bgp->import_check_table;
 	else
 		tree = &bgp->nexthop_cache_table;
+	if (json)
+		json_nexthops = json_object_new_array();
 	for (afi = AFI_IP; afi < AFI_MAX; afi++) {
-		frr_each (bgp_nexthop_cache, &(*tree)[afi], bnc)
-			bgp_show_nexthop(vty, bgp, bnc, false);
+		frr_each (bgp_nexthop_cache, &(*tree)[afi], bnc) {
+			if (json)
+				json_nexthop = json_object_new_object();
+			bgp_show_nexthop(vty, bgp, bnc, false, json_nexthop);
+			if (json)
+				json_object_array_add(json_nexthops,
+						      json_nexthop);
```

Comment:
In the nexthop array, would it be possible to make "ip" as the key for each entry? That would result in an output  like this:

{
  "default":{
    "nexthops":{
      **"6.0.0.28"**: { -----> "ip" is the key for this json object
        "valid":true,
        "complete":true,
        "igpMetric":0,
        "pathCount":40,
        "gate":"fe80::202:ff:fe00:1f",
        "ifname":"swp2",
        "lastUpdate":1666806289
      }
    }
}

Reply:
Fixed
```
leaf-A# show bgp nexthop 10.11.10.1 detail json 
{
  "nexthops":{
    "10.11.10.1":{
      "valid":true,
      "complete":true,
      "igpMetric":0,
      "pathCount":1,
      "peer":"10.11.10.1",
      "gates":[
        {
          "ifname":"eth1"
        }
      ],
      "lastUpdate":{
        "epoch":1667306862,
        "string":"Tue Nov  1 12:47:42 2022\n"
      },
      "paths":[
        {
          "afi":"IPv4",
          "safi":"unicast",
          "dest":"192.168.11.0/24",
          "vrf":"default",
          "flags":1168
        }
      ]
    }
  }
}
```

Reply:
Thank you for making the changes, this looks good. Could you please provide the output for "show bgp nexthops" as well?

Reply:
In the "paths" list can we change "dest" to "prefix" please?

Reply:
Renamed "dest" to "prefix" and here is also the output of "show bgp nexthop json"
```
leaf-A# show bgp nexthop 10.11.10.1 detail json
{
  "nexthops":{
    "10.11.10.1":{
      "valid":true,
      "complete":true,
      "igpMetric":0,
      "pathCount":1,
      "peer":"10.11.10.1",
      "gates":[
        {
          "ifname":"eth1"
        }
      ],
      "lastUpdate":{
        "epoch":1667413329,
        "string":"Wed Nov  2 18:22:09 2022\n"
      },
      "paths":[
        {
          "afi":"IPv4",
          "safi":"unicast",
          "prefix":"192.168.11.0/24",
          "vrf":"default",
          "flags":1168
        }
      ]
    }
  }
}
leaf-A# show bgp nexthop 
Current BGP nexthop cache:
 10.10.10.1 valid [IGP metric 0], #paths 1, peer 10.10.10.1
  if eth0
  Last update: Wed Nov  2 18:22:09 2022
 10.11.10.1 valid [IGP metric 0], #paths 1, peer 10.11.10.1
  if eth1
  Last update: Wed Nov  2 18:22:09 2022
leaf-A# show bgp nexthop json 
{
  "nexthops":{
    "10.10.10.1":{
      "valid":true,
      "complete":true,
      "igpMetric":0,
      "pathCount":1,
      "peer":"10.10.10.1",
      "gates":[
        {
          "ifname":"eth0"
        }
      ],
      "lastUpdate":{
        "epoch":1667413329,
        "string":"Wed Nov  2 18:22:09 2022\n"
      }
    },
    "10.11.10.1":{
      "valid":true,
      "complete":true,
      "igpMetric":0,
      "pathCount":1,
      "peer":"10.11.10.1",
      "gates":[
        {
          "ifname":"eth1"
        }
      ],
      "lastUpdate":{
        "epoch":1667413329,
        "string":"Wed Nov  2 18:22:09 2022\n"
      }
    }
  }
}
```